Fixes #477 - Dropped index only column in the model.

jzgithub1 opened the pull request:
Removed the requested change to the ModelBean and made all needed changes to webservice usages and ModelKeyParserTest.java

jzgithub1 commented:
@ivakegg please check this in the GUI because I don't have all the components of the GUI running right now.  Thanks.

jzgithub1 commented:
I think I have to get indexOnly out of json_Model.html too.

brianloss commented on warehouse/query-core/src/main/java/datawave/query/model/util/LoadModelFromXml.java:
```diff
@@ -46,9 +46,7 @@ public static QueryModel loadModelFromXml(InputStream stream) throws Exception {
-                    if (mapping.getIndexOnly()) {
-                        model.addUnevaluatedField(mapping.getFieldName());
-                    }
+                    model.addUnevaluatedField(mapping.getFieldName());
```
This line should go away, and in addition, the concept of unevaluated (aka index-only) fields should be removed from the QueryModel class entirely. In addition (question for @ivakegg), I'm wondering if we need to write a utility to convert an old format model into a new one (e.g., to re-write the model in the metadata table)?

jzgithub1 replied:
My QueryModel.java file is read-only for some reason.  Am I barred from editing it for some reason?  I could change the perms and edit it but I just want to see that I am not barred for some reason.

ivakegg replied:
@brianloss A conversion utility is not a bad thought.  We are trying to be backward compatible but it would be good to save the extra bits in the metadata table.  @jzgithub1 How about create a separate ticket for creating this utility please.

jzgithub1 replied:
Will create a ticket to write a utility to convert an old format model into a new one (e.g., to re-write the model in the metadata table).  

I checked for backward compatibility in the pull request by doing an insert model with the indexOnly attribute in the xml that I posted.  The insert still worked.

ivakegg commented on web-services/model/src/main/java/datawave/webservice/query/model/ModelKeyParser.java:
```diff
@@ -105,16 +102,14 @@ public static Mutation createDeleteMutation(FieldMapping mapping, String modelNa
             m = new Mutation(mapping.getFieldName());
-            m.putDelete(modelName + dataType, mapping.getModelFieldName() + indexOnly + NULL_BYTE + mapping.getDirection().getValue(), cv,
-                            System.currentTimeMillis());
+            m.putDelete(modelName + dataType, mapping.getModelFieldName() + NULL_BYTE + mapping.getDirection().getValue(), cv, System.currentTimeMillis());
```
We may need to add two delete mutations here: 1 for the original model entry and 1 for the new model entry.

brianloss replied:
We only need the two delete mutations (with and without index_only) in the forward section (the else block). The change in this section is fine.

brianloss replied:
> 
> 
> We only need the two delete mutations (with and without index_only) in the forward section (the else block). The change in this section is fine.

It doesn't appear this has been addressed yet.

jzgithub1 commented:
Bottom line up front: 
I am pretty sure we are fine with this one delete mutation after tracing the master branch code and new code while running the ModelBeanTest and ModelKeyParserTest.  The reason is because the same Mutation is created since indexOnly is false all the time which causes only empty string to be added to the column qualifier on line 111 of ModelKeyParser (in the call to mutation.putDelete) .  I am still evaluating this and should have a final comment by the end of day.  

I am tracing the function createDeleteMutation in the master branch code while running the ModelBeanTest.  If mapping.getIndexOnly() on line 108 in ModelKeyParser always evaluates to false (since it is never used as is written in the issue ticket)  then the 'indexOnly' string always evaluates to empty string.  So without indexOnly inside the field mapping the Mutation created is really the same as it always was before for legacy Models and will be in the future.  I did try to add a function called CreateDeleteMutationLegacyIndexOnly and got ModelKeyParser test to succeed.  I would call that function on line 452 in the ModelBean.java after the ModelKeyParser.createDeleteMutation call.  That way I could add both the Mutations to the batch writer inside of that loop surrounding line 452 that loops through the fields in the FieldMapping.  The ModelBeanTest.testModelDelete test would fail for no other reason than an improper number of calls to System.currentTimeMillis() because of easy mock. 

After doing the above I realized I should step back to what I have pushed already because indexOnly is always and always has been added to the delete Mutation as empty string.

jzgithub1 commented:
I have verified that insert and delete operations on the Model bean still work via the Poster REST client.  I have added the columnVisibility="VIS1" attribute (and removed the indexOnly atttibutes) to the test data in the '<datawave_proj>/warehouse/query-core/src/test/resources/models' directory. This attribute is required for inserts and deletes to succeed.   If the indexOnly attribute is still in the XML post, then the insert or delete will still work. The indexOnly attribute is vestigial now.  Verified the additions and deletions in the accumulo shell of course.  Need to give 'root' user the VIS1 authorization. I need to put this in the documentation.

jzgithub1 commented:
I am going to change the columnVisibility="VIS1" to columnVisibility="PUBLIC" in the test data.  Wait for the next commit please before merging.

jzgithub1 commented:
I am on vacation now Brian.   I tested this exhaustively with Poster application doing REST call for insert and delete of models and checking out the datawave.metadata table in the Accumulo shell.  It all works for me.  I will look at it again if you want next week when I come back.

brianloss commented:
> 
> 
> I am on vacation now Brian. I tested this exhaustively with Poster application doing REST call for insert and delete of models and checking out the datawave.metadata table in the Accumulo shell. It all works for me. I will look at it again if you want next week when I come back.

Yes, when you get back, the specific comment about 2 delete mutations (the comment I quoted) needs to be addressed. Everything will test correctly without the change, but it won't be backwards compatible with old data that had the index only marker.

jzgithub1 commented:
Thanks for the clarification.

On Tue, Jul 30, 2019 at 2:23 PM Brian Loss <notifications@github.com> wrote:

> I am on vacation now Brian. I tested this exhaustively with Poster
> application doing REST call for insert and delete of models and checking
> out the datawave.metadata table in the Accumulo shell. It all works for me.
> I will look at it again if you want next week when I come back.
>
> Yes, when you get back, the specific comment about 2 delete mutations (the
> comment I quoted) needs to be addressed. Everything will test correctly
> without the change, but it won't be backwards compatible with old data that
> had the index only marker.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NationalSecurityAgency/datawave/pull/488?email_source=notifications&email_token=ABNHZIXM3RH77UOBBRPVSULQCCBKNA5CNFSM4H4UTPPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3E3NNI#issuecomment-516535989>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABNHZIQPQWZWYPGEAEMYHQLQCCBKNANCNFSM4H4UTPPA>
> .
>

jzgithub1 commented:
Made the requested change @brianloss  and tested it.  Sorry for the delay on this.
